File: scala/scala-impl/src/org/jetbrains/plugins/scala/packagesearch/util/AsyncExpirableCache.scala
package org.jetbrains.plugins.scala.packagesearch.util

import com.github.benmanes.caffeine.cache.{AsyncLoadingCache, Caffeine}
import org.jetbrains.annotations.ApiStatus.Internal
import org.jetbrains.annotations.VisibleForTesting
import org.jetbrains.plugins.scala.packagesearch.util.AsyncExpirableCache.{ExpirationTimeout, MaxSize}

import java.time.Duration
import java.util.concurrent.{CompletableFuture, ExecutorService}

final class AsyncExpirableCache[K, V](maxSize: Long,
                                      expirationTimeout: Duration,
                                      executor: ExecutorService,
                                      computeValue: K => V) {
  def this(executor: ExecutorService, computeValue: K => V) =
    this(MaxSize, ExpirationTimeout, executor, computeValue)

  private[this] val cache: AsyncLoadingCache[K, V] = Caffeine.newBuilder()
    .maximumSize(maxSize)
    .expireAfterWrite(expirationTimeout)
    .executor(executor)
    .buildAsync(computeValue(_))

  def get(key: K): CompletableFuture[V] = cache.get(key)

  @Internal
  @VisibleForTesting
  def updateCache(key: K, value: CompletableFuture[V]): Unit = cache.put(key, value)
}

object AsyncExpirableCache {
  private val MaxSize = 10_000L
  private val ExpirationTimeout = Duration.ofMinutes(10)
}

File: scala/package-search-client/src/org/jetbrains/plugins/scala/packagesearch/api/AsyncExpirableCache.kt
package org.jetbrains.plugins.scala.packagesearch.api

import com.github.benmanes.caffeine.cache.Caffeine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.future.future
import org.jetbrains.annotations.ApiStatus
import org.jetbrains.annotations.VisibleForTesting
import java.util.concurrent.CompletableFuture
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.toJavaDuration

class AsyncExpirableCache<K, V>(
    cs: CoroutineScope,
    maxSize: Long = 10_000,
    expirationTimeout: Duration = 10.minutes,
    computeValue: suspend (K) -> V,
) {
    private val cache = Caffeine.newBuilder()
        .maximumSize(maxSize)
        .expireAfterWrite(expirationTimeout.toJavaDuration())
        .buildAsync<K, V> { key, _ -> cs.future { computeValue(key) } }

    fun get(key: K): CompletableFuture<V> = cache.get(key)

    @ApiStatus.Internal
    @VisibleForTesting
    internal fun updateCache(key: K, value: CompletableFuture<V>) = cache.put(key, value)
}
